create transform matrix from GCPS

vincentsarago opened the pull request:
ref #1479
👋 @sgillies,
This PR is a first sketch for getting the geotransform from GCPs. The code itself works but the logic might not be perfect. Right now what happens is:
- when opening a files rasterio will try to fetch de transform matrix using `GDALGetGeoTransform`
- if `GDALError.failure` is returned then it fetches the gcps (L308). (need better error handling) 
- after `self.meta` call it check if the CRS is defined, and if `_gcps` is set then it get the CRS from it <--- this is not great 

### To Do
- [x] better logic (especialy for CRS)
- [x] better error handling
- [x] add tests

I know this issue is set for rasterio 1.1, but someone asked me if it was possible to use rio-tiler on sentinel-1 data so I wanted to give it a try. ref: https://github.com/RemotePixel/remotepixel-tiler/issues/10. 

Feel free to reject/close this PR if it's adding noise to the repo 😄

sgillies commented on rasterio/_base.pyx:
```diff
@@ -303,9 +305,27 @@ cdef class DatasetBase(object):
+                        gcplist[i].dfGCPLine = obj.row
+                        gcplist[i].dfGCPX = obj.x
+                        gcplist[i].dfGCPY = obj.y
+                        gcplist[i].dfGCPZ = obj.z or 0.0
```
@vincentsarago maybe use just GDALGetGCPs if you can?

vincentsarago commented:
🤔 So the more I think about it, the more I think the implementation is 👎. With the current logic the file will get the transform matrix from the gcps anytime and this is maybe not expected from users. 
On GDAL gcps are used only when doing warping and with options (e.g: `-tps`). 

I guess on rasterio this should be the same, meaning we only use gcps information when doing tranformation and if (and only if) the user pass a certain option

vincentsarago commented:
@sgillies I've changed most of the original commit.
now the `transform` created from GCPS is added to `src.gcps` info. This make things almost seamless for the user (don't really change rasterio behaviour on file open) and enable the use of VRT like:

```python
with rasterio.open(src_path) as src_dst:
    src_crs = src_dst.gcps[1]
    src_transform = src_dst.gcps[2]
    with WarpedVRT(src_dst, src_crs=src_crs, crs=src_crs, src_transform=src_transform) as vrt_dst:
``` 

a more complex example can be found over https://gist.github.com/vincentsarago/076e674bdd0713890200e119bf2c055c

sgillies commented:
@vincentsarago I like this approach, though it can break some user code that unpacks the `gcps` object. That's my mistake in making it a tuple instead of an object.

Perhaps its time to turn it into an actual class using attrs (or not) with `.crs`, `.transform`, and `.points` properties? We could maintain backwards compatibility by providing the old indexing behavior: `points, crs = gcps`.

vincentsarago commented:
> Perhaps its time to turn it into an actual class using attrs (or not) with .crs, .transform, and .points properties? We could maintain backwards compatibility by providing the old indexing behavior: points, crs = gcps.

@sgillies it might seems a bit of a `hack` but I understand the need of  backwards compatibility

here is a sketch of a GCPS class
```python
    class GCPS(object):
        """Base for a GCPS class."""

        def __init__(self, points, crs, transform):
            """

            Parameters
            ----------
            """
            self._points = points
            self._crs = crs
            self._transform = transform

        def __iter__(self):
            return iter((self.points, self.crs))

        @property
        def points(self):
            return self._points

        @property
        def crs(self):
            return self._crs
        
        @property
        def transform(self):
            return self._transform
```

sgillies commented:
@vincentsarago yes, that's it.

For what it's worth, there's a different Python algorithm for fitting matrices to points at https://elonen.iki.fi/code/misc-notes/affine-fit/.

vincentsarago commented:
@sgillies This should be ready to be reviewed. As discussed I added a `GCPS` base class in `rasterio.control` which handle `points`, `crs` and `transform` information created from GCPS at file opening. This class also provide backward compatibility for unpacking:

```python
points, crs = src.gcps
points = src.gcps.points
crs = src.gcps.crs
transform = src.gcps.transform
```

I rebased this PR on maint-1.0 but I can change it back to master if needed

sgillies commented:
@vincentsarago my take is that we should do this for 1.1 and think about making it an excellent new feature instead of shoehorning it into an existing feature. Instead of only setting the transform when the GCPS instance is created by a dataset, it would be neat to have a general purpose GCP-to-transform function. For example:

```python
from rasterio.control import GroundControlPoint
from rasterio.transform import from_gcps

transform = from_gcps([GroundControlPoint(...), ...])
```

I'm getting bored with endless 1.0.x releases, it would be nice to get a new point release out there :)

vincentsarago commented:
@sgillies I love the idea of `transform.from_gcps` 👌 

I'll update this PR and remove the GCPS class.

sgillies commented on rasterio/_transform.pyx:
```diff
@@ -0,0 +1,38 @@
+    finally:
+            CPLFree(gcplist)
+
+    return [gt[i] for i in range(6)]
```
```suggestion
    return Affine.from_gdal([gt[i] for i in range(6)])
```

sgillies commented on rasterio/_transform.pyx:
```diff
@@ -0,0 +1,38 @@
+            gcplist[i].dfGCPY = obj.y
+            gcplist[i].dfGCPZ = obj.z or 0.0
+
+        err = GDALGCPsToGeoTransform(len(gcps), gcplist, gt, 0)
```
`&gt` I think?

vincentsarago replied:
@sgillies I have removed `&` before `gt` because it could not compile

sgillies commented on rasterio/_transform.pyx:
```diff
@@ -0,0 +1,38 @@
+                NotGeoreferencedWarning)
+
+    finally:
+            CPLFree(gcplist)
```
@vincentsarago right on! You're getting very good at Cython programming with GDAL :tophat:

vincentsarago commented:
@sgillies this is ready from my POV, let me know if it needs more tests or documentation

sgillies commented on rasterio/rio/info.py:
```diff
@@ -101,6 +102,8 @@ def info(ctx, input, aspect, indent, namespace, meta_member, verbose, bidx,
             else:
                 info['gcps']['crs'] = None
 
+            info['gcps']['transform'] = from_gcps(gcps)
```
:+1:

sgillies commented on rasterio/transform.py:
```diff
@@ -249,3 +250,8 @@ def rowcol(transform, xs, ys, op=math.floor, precision=None):
+
+
+def from_gcps(gcpslist):
+    """Return Affine transform created from GCPS."""
```
```suggestion
    """Make an Affine transform from ground control points

    Parameters
    ----------
    gcps : sequence of GroundControlPoint
        Such as the first item of a dataset's `gcps` property.
        
    Returns
    -------
    Affine

    """
```

sgillies commented on rasterio/transform.py:
```diff
@@ -249,3 +250,8 @@ def rowcol(transform, xs, ys, op=math.floor, precision=None):
     return rows, cols
+
+
+def from_gcps(gcpslist):
```
```suggestion
def from_gcps(gcps):
```

sgillies commented:
@vincentsarago I messed up and merged this to maint-1.0 by mistake and had to revert it. Can you rebase your gcptransform branch on master and submit a new PR? Sorry!

vincentsarago commented:
@sgillies sure but I'm not sure to follow why ? this is not a breaking change and IMO we could ship this is 1.0.29 ?
